slf4j-api/src/main/java/org/slf4j/helpers/BasicMDCAdapter.java

```diff
@@ -71,13 +79,14 @@ static boolean isJDK14() {
    * @throws IllegalArgumentException
    *                 in case the "key" parameter is null
    */
+  @Override
   public void put(String key, String val) {
     if (key == null) {
       throw new IllegalArgumentException("key cannot be null");
     }
-    Map<String, String> map = (Map<String, String>) inheritableThreadLocal.get();
+    Map<String, String> map = inheritableThreadLocal.get();
     if (map == null) {
-      map = Collections.<String, String>synchronizedMap(new HashMap<String, String>());
+      map = new HashMap<String, String>();
```

Comment:
Since every new thread inherits a copy of its parents map when the child thread is constructed.
Also, childValue method is called from within the parents thread during child thread construction.